src/relay/transforms/to_mixed_precision.cc

```diff
@@ -0,0 +1,420 @@
+    Array<Type> new_arg_types;
+
+    if (pre_call_node->op.as<FunctionNode>()) {
+      // Function Nodes don't store type info in the Call, it should be a []
+      new_arg_types = pre_call_node->type_args;
+    } else {
+      new_arg_types = call_args_and_types.second;
+    }
+
+    // Finally create the new attributes.
+    if (final_category == MIXED_PRECISION_ALWAYS) {
+      Attrs new_attrs = GetNewAttrs(pre_call_node, accumulation_dtype);
+      Expr output = Call(cur_op, new_args, new_attrs, new_arg_types, pre_call_node->span);
+      if (accumulation_dtype != output_dtype) {
+        output = CastArg(output, GetType(output), output_dtype);
+      }
+      return output;
+    }
+
+    return Call(cur_op, new_args, pre_call_node->attrs, new_arg_types, pre_call_node->span);
+  }
+
+  Expr VisitExpr_(const FunctionNode* func) final {
+    // Erase the ret_type annotation and let the normal pass recalculate
+    const_cast<FunctionNode*>(func)->ret_type = Type(nullptr);
+    return ExprMutator::VisitExpr_(func);
+  }
+
+  Expr VisitExpr_(const LetNode* op) final {
+    // First convert as much of the bound computation to lower precision as possible
+    Expr value = this->Mutate(op->value);
+
+    // Then rewrite the var type and associated expression
+    Var var = Downcast<Var>(this->Mutate(op->var));
+    VarNode* mutable_var = const_cast<VarNode*>((op->var).as<VarNode>());
+    mutable_var->type_annotation = GetType(value);
+    mutable_var->checked_type_ = mutable_var->type_annotation;
+
+    // Mutate body last as it may depend on previous results
+    Expr body = this->Mutate(op->body);
+    return Let(var, value, body, op->span);
+  }
+
+  // To access map of ops not registered for error reporting
+  friend Expr ToMixedPrecision(const Expr& expr, const DataType& mixed_precision_type,
+                               int missing_op_mode);
+};
+
+Expr ToMixedPrecision(const Expr& expr, const DataType& mixed_precision_type, int missing_op_mode) {
+  /*
+  missing_op_mode:
+
+  0: Does not allow any missing ops. Will throw errors and terminate the pass when encountering any.
+  1: Allow missing ops but throw warnings.
+  2: Allow missing ops and silently ignore them.
+  */
+  ICHECK(missing_op_mode >= 0 && missing_op_mode <= 2)
+      << " missing_op_mode must be either 0, 1, or 2 got " << missing_op_mode;
+
+  MixedPrecisionPass converter = MixedPrecisionPass(mixed_precision_type);
+  auto result = converter.Mutate(expr);
+
+  for (auto it = converter.missing_ops.begin();
+       missing_op_mode != 2 && it != converter.missing_ops.end(); it++) {
+    std::string op_name = it->first;
+    int appear_count = it->second;
+
+    LOG(WARNING) << "Op \"" << op_name << "\" not registered "
+                 << "FTVMMixedPrecisionConversionType appears " << appear_count << " in graph.";
+  }
+
+  if (converter.missing_ops.size() != 0 && missing_op_mode == 0) {
+    CHECK(0) << "Missing ops were found, please fix!";
```

Comment:
I think this is not supposed to be fixed by end-users. We should direct them to the discuss forum to post what is missing, or provide the tutorial link of AMP pass in the future to help them fix it. cc @areusch 
```suggestion
    CHECK(0) << "Missing ops were found";
```

Reply:
Ok so I guess my plan for this is make the suggested changes. Write a tutorial (update the RFC here https://github.com/apache/tvm-rfcs/pull/6 with the work to do) and then in the future link it.

Reply:
ive made the highlighted change